### PR TITLE
[LingBot World] Exact and fused fast paths for the streaming Wan VAE decoder (stacked on #7648, #7651)

### PR DESCRIPTION
**Pull request by tzhouam**
The streaming Wan VAE decoder spends substantial time on temporary buffers, layout conversions and separate normalization/activation launches. This change adds two opt-in decoder paths through `lingbot_vae_decode_fast_path`:

- `exact` reuses sharded convolution input buffers and removes the nearest-upsample fp32 round trip. The reusable installer can also pre-cast convolution parameters for callers using autocast; LingBot passes `conv_dtype=None` because its VAE already uses the stage dtype.
- `fused` additionally uses channels-last layouts, fused RMSNorm/SiLU kernels, a layout-preserving nearest upsample and temporal interleave. It supports both diffusers' `WanRMS_norm` and vLLM-Omni's `RMSNormVAE`, with eager fallbacks outside the supported kernel domain. The kernels are adapted from SGLang's Apache-2.0 diffusion kernel library.

The default remains off. `fused` is not bit-exact. The user accepted its visual quality on 2026-09-19 after reviewing the saved control/fused videos. Historical pre-encode streaming-decode diagnostics measured 54.4–55.1 dB PSNR; this is diagnostic evidence, not the quality acceptance criterion.

## Dependencies

Stacked on #7648 at `917715e1b` and #7651 at `956f56a1d`, combined base `5e6639ba3`. The fast-path commits are `0ad518c53`, `25204cac2`, `19beb9a97`, followed by the test-fixture fix `322922593`.

The rebase preserves selective padding initialization and automatic VAE sharding through `sequence_parallel_size`. Neither `vae_patch_parallel_size` nor `lingbot_vae_decode_dtype` is needed. Draft while the dependencies remain open.

## Served measurements after rebase

Revision `322922593`, 4xH200, 480x832, Ulysses SP-4, bf16, regional compile, WebSocket serving. Each arm uses one warmup plus three measured sessions of 40 chunks under the same GPU reservation.

| arm | steady median ms/chunk | RTF@16fps | post_decode ms |
|---|---:|---:|---:|
| control | 714 | 0.980 | 330 |
| exact | 713 | 0.979 | 328 |
| fused | 696 | 0.957 | 302 |
| repeated control | 715 | 0.981 | 330 |
| repeated fused | 685 | 0.945 | 300 |

Fused saves **18–30 ms/chunk** in these runs, with a consistent **28–30 ms reduction in post_decode**. The exact-only timing difference is below the run-to-run floor. The earlier 718 → 684–685 ms comparison was measured on the previous dependency stack and is not substituted for these results.

The fused runs produce byte-identical videos. Exact matches the repeated control byte-for-byte; the two controls differ from each other, so compiled video parity is not the lossless gate.

## Validation

- All pre-commit checks pass for the five changed files.
- Combined decoder, spatial-shard and pipeline CPU tests: **207 passed, 3 GPU tests deselected**. The decoder fixture now handles norms already patched during pipeline test collection.
- All five served timing arms completed. Startup logs verify automatic sharding and the requested exact/fused installation on every rank.
- Separate served eager validation: **40/40 measured chunks have identical raw decoder input and output hashes on each of four ranks**. The saved eager videos are also byte-identical. This correctness-only run uses one warmup and one measured session; its extra hashing/synchronization is excluded from timing comparisons.

### CODE REVIEW

File: vllm_omni/diffusion/distributed/autoencoders/wan_decoder_fast_path.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+logger = init_logger(__name__)
+
+_INSTALLED_ATTR = "_vllm_omni_wan_decoder_fast_path"
+LEVELS = ("exact", "fused")
+# diffusers' WanRMS_norm, and vLLM-Omni's RMSNormVAE that patch_wan_rms_norm swaps in for every Wan-family VAE.
+_NORM_CLASSES = ("WanRMS_norm", "RMSNormVAE")
+_CACHE_T = 2  # diffusers.models.autoencoders.autoencoder_kl_wan.CACHE_T
+
+
+def _upsample_forward(self: nn.Upsample, x: torch.Tensor) -> torch.Tensor:
```
Comment (Gaohan123) [inline vllm_omni/diffusion/distributed/autoencoders/wan_decoder_fast_path.py]: Please move these functions into utils

File: vllm_omni/diffusion/distributed/autoencoders/wan_spatial_shard.py [hunk 2]
```diff
@@ -240,8 +240,31 @@ def gather_and_trim_extent(
     return out
 
 
+def _persistent_input_buffer(module: nn.Module, shape: tuple[int, ...], reference: torch.Tensor) -> torch.Tensor:
```
Comment (Gaohan123) [inline vllm_omni/diffusion/distributed/autoencoders/wan_spatial_shard.py]: The same

File: vllm_omni/diffusion/models/lingbot_world/pipeline.py [hunk 3]
```diff
@@ -351,6 +367,27 @@ def _decode_source_image_fingerprinted(fingerprint: tuple[str, int, int, int, in
     return _decode_source_image_file(fingerprint[0])
 
 
+def _vae_decode_fast_path(model_config: dict) -> str | None:
```
Comment (Gaohan123) [inline vllm_omni/diffusion/models/lingbot_world/pipeline.py]: The same

File: vllm_omni/diffusion/distributed/autoencoders/wan_decoder_fast_path.py (hunk 1 above)
Comment (tzhouam) [reply vllm_omni/diffusion/distributed/autoencoders/wan_decoder_fast_path.py]: Moved the standalone upsample, layout, and resample helpers into `wan_decoder_utils.py`. The fast path module retains the fused norm module and installer. The focused decoder suite passes (11 tests, including two-rank cases); Ruff and formatting checks pass.

File: vllm_omni/diffusion/distributed/autoencoders/wan_spatial_shard.py (hunk 2 above)
Comment (tzhouam) [reply vllm_omni/diffusion/distributed/autoencoders/wan_spatial_shard.py]: Moved `_persistent_input_buffer` into `wan_decoder_utils.py` and imported it from the spatial shard module. The focused decoder suite passes (11 tests, including the two-rank buffer tests).

File: vllm_omni/diffusion/models/lingbot_world/pipeline.py (hunk 3 above)
Comment (tzhouam) [reply vllm_omni/diffusion/models/lingbot_world/pipeline.py]: Moved `_vae_decode_fast_path` into `lingbot_world/utils.py` and imported it in the pipeline. Ruff and formatting checks pass.

**Comment by tzhouam** [comment]
Author self-review (head `2023ef8a8`):

- Checked the opt-in `exact` and `fused` installation paths, the persistent input buffer and halo assembly, the fallback paths for unsupported fused-kernel inputs, and the planar RGB stride-specific copy.
- Addressed the three inline organization comments by moving standalone decoder helpers, the persistent buffer helper, and the LingBot fast-path config parser into utility modules. This commit only moves code and updates imports.
- Reran `tests/diffusion/distributed/test_wan_decoder_fast_path.py` in the compatible project environment: 11 passed, including two-rank cases. Ruff check, format check, and `git diff --check` pass. I did not rerun the H200 benchmark for this code-motion commit.

For the performance questions: the decoder kernel ledger in the PR description attributes about 91 ms GPU time per chunk per rank, with roughly 2,300 glue launches. The reported 4×H200, 480×832, four-step workload improves mean chunk-6–38 cadence from 590.94 to 548.68 ms with the combined CPU-layout and fused-decoder arm. That is a combined end-to-end result; I have not supplied a same-head ablation that isolates each optimization, so I do not claim an individual delta for either one. The `exact` path has bit-identical CPU and two-rank tests; the `fused` path is opt-in and still needs production visual approval.

**Review by Gaohan123 — APPROVED** [review]
LGTM. Thanks
